core/src/main/java/edu/wpi/grip/core/metrics/BenchmarkRunner.java

```diff
@@ -0,0 +1,131 @@
+package edu.wpi.grip.core.metrics;
+
+import edu.wpi.grip.core.events.BenchmarkEvent;
+import edu.wpi.grip.core.events.RunStartedEvent;
+import edu.wpi.grip.core.events.RunStoppedEvent;
+import edu.wpi.grip.core.events.StartSingleBenchmarkRunEvent;
+import edu.wpi.grip.core.events.TimerEvent;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Benchmark runner. This runs the pipeline multiple times to collect data about how long each
+ * step takes to run.
+ */
+@Singleton
+public class BenchmarkRunner {
+
+  private final EventBus eventBus;
+
+  /**
+   * State flag.
+   */
+  private final AtomicBoolean isBenchmarking = new AtomicBoolean(false);
+
+  /**
+   * The number of runs left in the benchmark.
+   */
+  private final AtomicInteger runsRemaining = new AtomicInteger(0);
+
+  /**
+   * Keep track of timers so we can reset them after the benchmark finishes running.
+   */
+  private final AtomicReference<Set<Timer>> timers = new AtomicReference<>(new HashSet<>());
+
+  @Inject
+  BenchmarkRunner(EventBus eventBus) {
+    this.eventBus = eventBus;
+  }
+
+  /**
+   * Benchmarks the pipeline.
+   *
+   * @param numRuns the number of runs of the pipeline to collect events from.
+   */
+  public void run(int numRuns) {
+    checkArgument(numRuns > 0, "Must run the pipeline at least once");
+    this.isBenchmarking.set(true);
+    this.runsRemaining.set(numRuns);
+    eventBus.post(BenchmarkEvent.started());
+    eventBus.post(new StartSingleBenchmarkRunEvent());
+  }
+
+  @Subscribe
+  @SuppressWarnings("PMD.UnusedPrivateMethod")
+  private void onTimerEvent(TimerEvent event) {
+    this.timers.get().add(event.getTimer());
+  }
+
+  @Subscribe
+  @SuppressWarnings({"PMD.UnusedPrivateMethod", "PMD.UnusedFormalParameter"})
+  private void onRunStart(@Nullable RunStartedEvent event) {
```

Comment:
No, it's reflective access so the event bus makes the method accessible when adding it to the subscribers list.